src/Symmetric/Openssl.php

```diff
@@ -372,6 +397,86 @@ public function getPadding()
     }
 
     /**
+     * Set Additional Authentication Data
+     *
+     * @param string $aad
+     * @return self
+     */
+    public function setAad($aad)
+    {
+        if (! $this->isAuthEncAvailable()) {
+            throw new Exception\RuntimeException(
+                'You need PHP 7.1+ and OpenSSL with CCM or GCM mode to use AAD'
+            );
+        }
+        if ($this->getMode() !== 'gcm' && $this->getMode() !== 'ccm') {
+            throw new Exception\RuntimeException(
+                'You can set the Tag Size only for CCM or GCM mode'
+            );
+        }
+        $this->aad = $aad;
+        return $this;
+    }
+
+    /**
+     * Get the Additional Authentication Data
+     *
+     * @return string
+     */
+    public function getAad()
+    {
+        return $this->aad;
+    }
+
+    /**
+     * Get the authentication tag
+     *
+     * @return string
+     */
+    public function getTag()
+    {
+        return $this->tag;
+    }
+
+    /**
+     * Set the tag size for CCM and GCM mode
+     *
+     * @param int $size
+     * @return self
+     */
+    public function setTagSize($size)
+    {
+        $size = (int) $size;
+        if (! $this->isAuthEncAvailable()) {
+            throw new Exception\RuntimeException(
+                'You need PHP 7.1+ and OpenSSL with CCM or GCM mode to set the Tag Size'
+            );
+        }
+        if ($this->getMode() !== 'gcm' && $this->getMode() !== 'ccm') {
+            throw new Exception\RuntimeException(
+                'You can set the Tag Size only for CCM or GCM mode'
+            );
+        }
```

Comment:
These previous two blocks are repeated in `setAad()`; perhaps you could extract those validations to a separate method?